cluster/src/shard_lock_manager.rs

```diff
@@ -0,0 +1,505 @@
+        let (mut keeper, mut stream) = etcd_client
+            .lease_keep_alive(self.id)
+            .await
+            .context(KeepAlive)?;
+        let new_expired_at = Self::keep_alive_once(&mut keeper, &mut stream).await?;
+        self.expired_at.store(new_expired_at, Ordering::Relaxed);
+
+        // send keepalive request every ttl/3.
+        let keep_alive_interval = self.ttl / 3;
+        let lease_id = self.id;
+        let expired_at = self.expired_at.clone();
+        let handle = runtime.spawn(async move {
+            loop {
+                match Self::keep_alive_once(&mut keeper, &mut stream).await {
+                    Ok(new_expired_at) => {
+                        debug!("The lease {lease_id} has been kept alive, new expire time in milliseconds:{new_expired_at}");
+                        expired_at.store(new_expired_at, Ordering::Relaxed);
+                    }
+                    Err(e) => {
+                        error!("Failed to keep lease alive, id:{lease_id}, err:{e}");
+                        if notifier.send(Err(e)).is_err() {
+                            error!("failed to send keepalive failure, lease_id:{lease_id}");
+                        }
+
+                        return
+                    }
+                }
+                let sleeper = tokio::time::sleep(keep_alive_interval);
+                tokio::select! {
+                    _ = sleeper => {
+                        debug!("Try to keep the lease alive again, id:{lease_id}");
+                    },
+                    _ = &mut stop_receiver => {
+                        debug!("Stop keeping lease alive, id:{lease_id}");
+                        if notifier.send(Ok(())).is_err() {
+                            error!("failed to send keepalive stopping message, lease_id:{lease_id}");
+                        }
+                        return
+                    }
+                }
+            }
+        });
+
+        self.keep_alive_handle = Some(handle);
+
+        Ok(())
+    }
+}
+
+/// Lock for a shard.
+///
+/// The lock is a temporary key in etcd, which is created with a lease.
+pub struct ShardLock {
+    shard_id: ShardId,
+    /// The temporary key in etcd
+    key: Bytes,
+    /// The value of the key in etcd
+    value: Bytes,
+    /// The lease of the lock in etcd
+    ttl_sec: u64,
+
+    lease_id: Option<i64>,
+    lease_check_handle: Option<JoinHandle<()>>,
+    lease_keepalive_stopper: Option<oneshot::Sender<()>>,
+}
+
+impl ShardLock {
+    fn new(shard_id: ShardId, key_prefix: &str, value: Bytes, ttl_sec: u64) -> Self {
+        Self {
+            shard_id,
+            key: Self::lock_key(key_prefix, shard_id),
+            value,
+            ttl_sec,
+
+            lease_id: None,
+            lease_check_handle: None,
+            lease_keepalive_stopper: None,
+        }
+    }
+
+    fn lock_key(key_prefix: &str, shard_id: ShardId) -> Bytes {
+        let key = format!("{key_prefix}/{shard_id:0>20}");
```

Comment:
Only after see your tests I know `0>20` meaning, better to leave a comment here.